src/libraries/System.Private.CoreLib/src/System/Buffers/Binary/BinaryPrimitives.ReverseEndianness.cs

```diff
@@ -0,0 +1,268 @@
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static long Reverse(long value) => ReverseEndianness(value);
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector128<long> Reverse(Vector128<long> vector) =>
+                Vector128.Shuffle(vector.AsByte(), Vector128.Create((byte)7, 6, 5, 4, 3, 2, 1, 0, 15, 14, 13, 12, 11, 10, 9, 8)).AsInt64();
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector256<long> Reverse(Vector256<long> vector) =>
+                Vector256.Shuffle(vector.AsByte(), Vector256.Create((byte)7, 6, 5, 4, 3, 2, 1, 0, 15, 14, 13, 12, 11, 10, 9, 8, 23, 22, 21, 20, 19, 18, 17, 16, 31, 30, 29, 28, 27, 26, 25, 24)).AsInt64();
+        }
+
+        private static void ReverseEndianness<T, TReverser>(ReadOnlySpan<T> source, Span<T> destination)
+            where T : struct
+            where TReverser : IEndiannessReverser<T>
+        {
+            if (destination.Length < source.Length)
+            {
+                ThrowDestinationTooSmall();
+            }
+
+            ref T sourceRef = ref MemoryMarshal.GetReference(source);
+            ref T destRef = ref MemoryMarshal.GetReference(destination);
+            ref T end = ref Unsafe.Add(ref sourceRef, source.Length);
+
+            if (Unsafe.AreSame(ref sourceRef, ref destRef) ||
+                !source.Overlaps(destination, out int elementOffset) ||
+                elementOffset < 0)
+            {
+                // Either there's no overlap between the source and the destination, or there's overlap but the
+                // destination starts at or before the source.  That means we can safely iterate from beginning
+                // to end of the source and not have to worry about writing into the destination and clobbering
+                // source data we haven't yet read.
+
+                if (Vector128.IsHardwareAccelerated && source.Length >= Vector128<T>.Count)
+                {
+                    if (Vector256.IsHardwareAccelerated && source.Length >= Vector256<T>.Count)
+                    {
+                        ref T end256 = ref Unsafe.Add(ref sourceRef, source.Length - Vector256<T>.Count + 1);
+                        do
+                        {
+                            Vector256.StoreUnsafe(TReverser.Reverse(Vector256.LoadUnsafe(ref sourceRef)), ref destRef);
+                            sourceRef = ref Unsafe.Add(ref sourceRef, Vector256<T>.Count);
+                            destRef = ref Unsafe.Add(ref destRef, Vector256<T>.Count);
+                        }
+                        while (Unsafe.IsAddressLessThan(ref sourceRef, ref end256));
+                    }
+                    else
+                    {
+                        ref T end128 = ref Unsafe.Add(ref sourceRef, source.Length - (Vector128<T>.Count - 1));
```

Comment:
Nit:
```suggestion
                        ref T end128 = ref Unsafe.Add(ref sourceRef, source.Length - Vector128<T>.Count + 1);
```
To be the same as above for `end256`. Or vice-versa.

BTW: should there be `Unsafe.IsAddressLessThanOrEqual` to avoid the +1 for comparison.

Reply:
> Unsafe.IsAddressLessThanOrEqual

That doesn't exist. I'd be fine with us adding one, though.

Reply:
That's what I meant.